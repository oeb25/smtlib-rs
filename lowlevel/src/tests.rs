use crate::{ast::Script, lexicon::SmtlibParse, parse::Parser};

#[test]
fn escaped_string() {
    insta::assert_ron_snapshot!(String::parse(&mut Parser::new(
        r#""Hello ""world"" this is cool!""#
    )));
}

#[test]
fn bubble_sort() {
    insta::assert_ron_snapshot!(Script::parse(include_str!("../examples/bubble_sort.smt2")));
}

mod z3 {
    use crate::{Driver, ast::Command, backend::z3_binary::Z3Binary};

    macro_rules! cmd {
        ($d:expr, $cmd:literal) => {
            insta::assert_ron_snapshot!($d.exec(&Command::parse($cmd)?)?);
        };
    }

    #[test]
    fn echo_test() -> Result<(), Box<dyn std::error::Error>> {
        let mut d = Driver::new(Z3Binary::new("z3")?)?;

        cmd!(d, r#"(echo "Hello, world!")"#);
        cmd!(d, r#"(echo "Hello, unmatched paren! :)")"#);

        Ok(())
    }

    #[test]
    fn negative_numbers() -> Result<(), Box<dyn std::error::Error>> {
        let mut d = Driver::new(Z3Binary::new("z3")?)?;

        cmd!(d, r#"(declare-const x Int)"#);
        cmd!(d, r#"(assert (< x (- 1)))"#);
        cmd!(d, r#"(check-sat)"#);
        cmd!(d, r#"(get-model)"#);

        Ok(())
    }
}

#[cfg(feature = "z3-static")]
mod z3_static {
    use crate::{Driver, ast::Command, backend::Z3Static};

    macro_rules! cmd {
        ($d:expr, $cmd:literal) => {
            insta::assert_ron_snapshot!($d.exec(&Command::parse($cmd)?)?);
        };
    }

    #[test]
    fn echo_test() -> Result<(), Box<dyn std::error::Error>> {
        let mut d = Driver::new(Z3Static::new()?)?;

        cmd!(d, r#"(echo "Hello, world!")"#);
        cmd!(d, r#"(echo "Hello, unmatched paren! :)")"#);

        Ok(())
    }
}
