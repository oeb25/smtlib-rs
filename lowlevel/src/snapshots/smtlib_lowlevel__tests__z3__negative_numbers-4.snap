---
source: lowlevel/src/tests.rs
expression: "d.exec(Command :: parse(& st, r#\"(get-model)\"#)?)?"
---
SpecificSuccessResponse(GetModelResponse(GetModelResponse([
  DefineFun(FunctionDef(Symbol("x"), [], Sort(Simple(Symbol("Int"))), Application(Identifier(Simple(Symbol("-"))), [
    SpecConstant(Numeral(Numeral(
      repr: Number(2),
    ))),
  ]))),
])))
